### ISSUE TITLE
[Feature Request]: Get heap statistics for all VM instances

### ISSUE THREAD

**Issue by fabiospampinato** [electron/electron#27375]
**Preflight Checklist**

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

**Problem Description**

Essentially I'd like to be able to fetch all the data necessary to recreate this list available under the "Memory" tab in the devtools:

<img width="436" alt="Screen Shot 2021-01-19 at 21 08 08" src="https://user-images.githubusercontent.com/1812093/105093515-c61a2e00-5a9a-11eb-8647-2ba6cf4fc433.png">

**Proposed Solution**

Maybe there should be a `process.getHeapsStatistics` function or something?

**Alternatives Considered**

- Communicating with each worker manually to get this info? I'm not sure that even works if node integration is disabled in them.
- Not bothering about collecting memory usage statistics? It'd be useful to be able to track this automatically and over time via telemetry.

**Additional Information**

_No response_

**Comment by fabiospampinato** [comment]
Does anyone know how to get that information from the [debugger protocol thing](https://chromedevtools.github.io/devtools-protocol)? I'm thinking there should be a method there for retrieving that information, but I'm not finding it 🤔

**Comment by nornagon** [comment]
https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-getHeapUsage

**Comment by fabiospampinato** [comment]
@nornagon I've done some experimentation with that today, that gives you the numbers for the isolate (?) you are attached to only, not for all workers, in order to get the numbers for all isolates you need to:

1. Either attach to the debugger immediately, enable targets discovery immediately with `Target.setDiscoverTargets`, listen to events, keep track of the currently active workers manually by listening and reacting to the `Target.targetCreated`, `Target.targetDestroyed` and `Target.targetInfoChanged` events, then attach to each worker and send the `Runtime.getHeapUsage` command in each of them.

2. Or, it seems that by attaching to the debugger via `browserWindow#webContents#debugger` one is attaching to a "page", not to the "browser", which I think is probably the reason why the `Target.getTargets` command doesn't list workers at all and you need to keep track of them manually. If one could attach to the "browser" then there would be no need to keep track of workers manually and the whole thing would be simpler.

Long story short for saying: keeping track of memory usage via telemetry is probably something that more apps should do, so I'd argue this should be encouraged and made easy, the current ways of keeping track of memory usage is workers, unless I'm missing somethings, are waaaay too convoluted.

**Comment by nornagon** [comment]
@fabiospampinato there exists [`app.getAppMetrics()`](https://www.electronjs.org/docs/latest/api/app/#appgetappmetrics), which can tell you overall memory usage for each process. In some ways this is a better thing to monitor as it will give you the _total_ memory usage, including all threads, as opposed to only V8 heap usage, but it won't break it down by isolate. There is also [`process.getHeapStatistics()`](https://www.electronjs.org/docs/api/process#processgetheapstatistics), along with a related family of methods, which produce more detailed information. You'd need to have `nodeIntegrationInWorker` set to use those methods in workers, though.

I sketched out a PR that pulls this from roughly the same API that the devtools hits, but I'm not convinced yet that we should merge such an API. #30383

**Comment by fabiospampinato** [comment]
@nornagon Thanks for putting the PR together! I'm already using `app.getAppMetrics`, but as you point out that doesn't break down numbers by workers. `process.getHeapStatistics` would work for my use case by in order to use it I would need to have node integration enabled, as you mentioned, which I'd rather not have, and even with that enabled I would need to add a bit of code to each worker, and not only workers shouldn't need to know about this but some workers I'm using come from third-party libraries which I can't quite patch just for this.

IMO this is a relatively minor feature, but workers memory usage can get out of hand, if there was an API that I could use to easily keep track of this I would totally use it. I'm currently using the devtools for doing that locally get I can't quite use that to gather some telemetry about this.
